procrustes/psdp.py

```diff
@@ -33,10 +33,138 @@
 
 __all__ = [
     "psdp_woodgate",
-    "psdp_peng"
+    "psdp_peng",
+    "psdp_opt"
 ]
 
 
+def psdp_opt(
+    a: np.ndarray,
+    b: np.ndarray,
+    pad: bool = True,
+    translate: bool = False,
+    scale: bool = False,
+    unpad_col: bool = False,
+    unpad_row: bool = False,
+    check_finite: bool = True,
+    weight: Optional[np.ndarray] = None,
+) -> ProcrustesResult:
+    r"""
+    Spectral projected gradient method for the positive semi-definite Procrustes problem.
+    """
+    a, b = setup_input_arrays(
+        a,
+        b,
+        unpad_col,
+        unpad_row,
+        pad,
+        translate,
+        scale,
+        check_finite,
+        weight,
+    )
+    if a.shape != b.shape:
+        raise ValueError(
+            f"Shape of A and B does not match: {a.shape} != {b.shape} "
+            "Check pad, unpad_col, and unpad_row arguments."
+        )
+
+    # Initializing the required minimizer.
+    # Here, a denote the matrix to be transformed i.e. A, x is the transformer X and b is the
+    # target matrix B. Our goal is to minimize ||XA - B||_F (as mentioned in the function description).
+    n, m = a.shape
+    x = np.eye(n)
+
+    # Option structure with fields that serve as parameters for the algorithm.
+    opts = {
+        # Maximum number of iterations.
+        "mxitr": 10000,
+        # Stop control for ||X_k - X_{k-1}||_F.
+        "xtol": 1e-5,
+        # Stop control for |F_k - F_{k-1}|/(1+|F_{k-1}|).
+        "ftol": 1e-12,
+        # If proj is True we perform Cholesky decomposition else we do spectral
+        # decomposition.
+        "proj": True,
+        # Check if A is a diagonal matrix.
+        "a_is_diag": n == m and np.count_nonzero(a - np.diag(np.diagonal(a))) == 0,
+        # Parameter of the non-monotone technique proposed by Zhang-Hager.
+        "gamma": 0.85,
+        # Parameter for control the linear approximation in line search.
+        "rho": 1e-4,
+        # Factor for decreasing the step size in the backtracking line search.
+        "eta": 0.1,
+        # Initial step size.
+        "tau": 1e-3,
+    }
+
+    hold = np.dot(x, a) - b
+    grad = np.dot(hold, a.T)
+    f = np.linalg.norm(hold) ** 2
+    norm_grad = np.linalg.norm(grad)
+    q = 1
+    cval = f
+
+    # Main iteration of the algorithm.
+    for i in range(opts["mxitr"]):
+        x_old = x
+        f_old = f
+        grad_old = grad
+        derivative = opts["rho"] * (norm_grad**2)
+        nls = 1
+
+        while True:
+            x = x_old - opts["tau"] * grad_old
+            x = _psd_proj(x, opts["proj"])
+            hold = np.dot(x, a) - b
+            f = np.linalg.norm(hold) ** 2
+
+            if f <= cval - opts["tau"] * derivative or nls >= 5:
```

Comment:
I would guess `nls` is counting a number of iterations for the while loop. We should add this to the parameter list to allow users to specify the values they want.